### PR TITLE
PV2-539 - addition of new proxy service and service for LevyFunded pa…

### PR DESCRIPTION
**Pull request by davehowlett**
…yments to FundingSource.ServiceFabric application

### CODE REVIEW

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            IPaymentLogger paymentLogger,
+            IExecutionContext executionContext)
+        {
+            this.apprenticeshipKeyService = apprenticeshipKeyService;
+            this.proxyFactory = proxyFactory ?? new ActorProxyFactory();
+            this.paymentLogger = paymentLogger;
+            this.executionContext = (ESFA.DC.Logging.ExecutionContext)executionContext;
+        }
+
+        public async Task Handle(EarningEvent message, IMessageHandlerContext context)
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: Is this in the right place? Funding source app is handling required payments and producing funding source events, what is done here is a job of required payments app.

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces.csproj [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces.csproj]: Interface projects can be Netstandard, helps if we need to call them from tests

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            paymentLogger.LogInfo($"Processing LevyFundedProxyService event. Message Id : {context.MessageId}");
+            executionContext.JobId = message.JobId.ToString();
+
+            try
+            {
+                var key = apprenticeshipKeyService.GenerateApprenticeshipKey(
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: Actors will be employer scoped, not apprenticeship, this is not needed here

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/ILevyFundedService.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    /// <summary>
+    /// This interface defines the methods exposed by an actor.
+    /// Clients use this interface to interact with the actor that implements it.
+    /// </summary>
+    public interface ILevyFundedService : IActor
+    {
+        Task Initialise();
+
+        Task Reset();
+        Task<ReadOnlyCollection<RequiredPaymentEvent>> HandleEarningEvent(EarningEvent message, CancellationToken none);
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/ILevyFundedService.cs]: this needs to be handling required payments

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        public LevyFundedService(ActorService actorService, ActorId actorId,
+            IPaymentLogger paymentLogger,
+            IApprenticeshipKeyService apprenticeshipKeyService,
+            IPaymentKeyService paymentKeyService,
+        ILifetimeScope lifetimeScope) : base(actorService, actorId)
+        {
+            this.paymentLogger = paymentLogger;
+            this.paymentKeyService = paymentKeyService ?? throw new ArgumentNullException(nameof(paymentKeyService));
+            this.lifetimeScope = lifetimeScope;
+            apprenticeshipKey = apprenticeshipKeyService.ParseApprenticeshipKey(actorId.GetStringId());
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs]: This actor will be employer scoped, no need for apprenticeship key

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/ILevyFundedService.cs (hunk 4 above)
Comment (davehowlett) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/ILevyFundedService.cs]: of course! I did copy copy pasting on the implementations and then promptly forgot to change it!

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces.csproj (hunk 2 above)
Comment (davehowlett) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces/SFA.DAS.Payments.FundingSource.LevyFundedService.Interfaces.csproj]: done

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs (hunk 5 above)
Comment (davehowlett) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs]: done

File: src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs (hunk 1 above)
Comment (davehowlett) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: very good point - copy pasted from required payments and then submitted the PR before actually changing anything!

Comment (kuzinator) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: looks better, can we get rid of `apprenticeshipKeyService` and rename `requiredPaymentEvent` as a last thing here?

Comment (davehowlett) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: done

Comment (kuzinator) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedProxyService/Handlers/PayableEarningHandler.cs]: `requiredPaymentEvent` still there, the result contains funding source events, not required payments.

**Comment by kuzinator** [comment]
The build gives 50+ warnings on conflicting dependencies and proc architecture. We have agreed before to treat warnings as errors. Please tidy up.

**Pull request merged**